properties/extras/uid.py

```diff
@@ -0,0 +1,234 @@
+            return cls._INSTANCES[uid]
+        elif uid in cls._INSTANCES:
+            raise ValueError('Uid already used: {}'.format(uid))
+        elif uid not in registry:
+            raise ValueError('Invalid uid: {}'.format(uid))
+        value = registry[uid]
+        if not isinstance(value, HasUID):
+            try:
+                input_class = value.get('__class__')
+            except AttributeError:
+                input_class = None
+            new_cls = cls._deserialize_class(input_class, trusted, strict)
+            new_inst = new_cls()
+            registry.update({uid: new_inst})
+            super(HasUID, cls).deserialize(
+                value=value,
+                trusted=trusted,
+                strict=strict,
+                registry=registry,
+                _instance=new_inst,
+                **kwargs
+            )
+        cls._INSTANCES[uid] = registry[uid]
+        return registry[uid]
+
+
+class Pointer(base.Instance):
+    """Property for HasUID instances where string UID pointer may be used
+
+    **Available keywords** (in addition to those inherited from
+    :ref:`Instance <instance>`):
+
+    * **enforce_uid** - Require Pointer strings to resolve into instances.
+      If False, the default, the Pointer property may be set to an arbitrary
+      string; if True, the string must correspond to an existing UID.
+    * **uid_prop** - Property name of the UID property on instance_class.
+      The default is 'uid'.
+    """
+
+    class_info = 'an instance or uid of an instance'
+
+    @property
+    def enforce_uid(self):
+        """Require Pointer strings to resolve into instances
+
+        If False, the default, the Pointer may be set to an arbitrary string;
+        if True, the string must correspond to an existing UID.
+        """
+        return getattr(self, '_enforce_uid', False)
+
+    @enforce_uid.setter
+    def enforce_uid(self, value):
+        self._enforce_uid = bool(value)
+
+    @property
+    def uid_prop(self):
+        """Property name of the UID property on instance_class
+
+        The default is 'uid'
+        """
+        return getattr(self, '_uid_prop', 'uid')
+
+    @uid_prop.setter
+    def uid_prop(self, value):
+        self._uid_prop = text_type(value)
+
+    @property
+    def info(self):
+        info = '{} or a valid {} property of that class'.format(
+            super(Pointer, self).info, self.uid_prop
+        )
+        return info
+
+    def validate(self, instance, value):
+        instance_value = None
+        try:
+            if isinstance(value, string_types):
+                prop = getattr(
+                    self.instance_class, '_props', {}
+                ).get(self.uid_prop)
+                if prop:
+                    value = prop.validate(None, value)
+                if hasattr(self.instance_class, 'validate_uid'):
+                    self.instance_class.validate_uid(value)
+                if hasattr(self.instance_class, 'get_by_uid'):
```

Comment:
I am wondering if in between these two we can add a hook for "does this uid exist", which doesn't expect to receive an instance. This would be potentially really useful in cases when it is much cheaper to check if something exists than it is to retrieve it / load it into memory. For example, HDF5 files with big arrays, or DB-backed classes w/ quick lookups and long loads.

Reply:
I _think_ that would fall under `validate_uid`?

Reply:
Yeah, okay, that is fair. So I guess my concern here is that we are running validation on the whole tree by loading the instances and validating those. That could be _really_ expensive to do. Is there an in-between behaviour where we _only_ check that they're present, but not that the underlying data in the object pointed to is valid?